crates/uv-resolver/src/exclude_newer.rs

```diff
@@ -425,47 +430,172 @@ impl std::fmt::Display for ExcludeNewerValue {
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for PackageExcludeNewer {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct Visitor;
+
+        impl serde::de::Visitor<'_> for Visitor {
+            type Value = PackageExcludeNewer;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("a timestamp string or false/null to disable exclude-newer")
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                ExcludeNewerValue::from_str(v)
+                    .map(|ts| PackageExcludeNewer::Enabled(Box::new(ts)))
+                    .map_err(|e| E::custom(format!("failed to parse timestamp: {e}")))
+            }
+
+            fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                if v {
+                    Err(E::custom(
+                        "expected false to disable exclude-newer, got true",
+                    ))
+                } else {
+                    Ok(PackageExcludeNewer::Disabled)
+                }
+            }
+
+            fn visit_none<E>(self) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(PackageExcludeNewer::Disabled)
+            }
+
+            fn visit_unit<E>(self) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(PackageExcludeNewer::Disabled)
+            }
+        }
+
+        deserializer.deserialize_any(Visitor)
+    }
+}
+
+impl serde::Serialize for PackageExcludeNewer {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::Enabled(timestamp) => timestamp.to_string().serialize(serializer),
+            Self::Disabled => serializer.serialize_bool(false),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PackageExcludeNewerChange {
+    Disabled { was: ExcludeNewerValue },
+    Enabled { now: ExcludeNewerValue },
+    TimestampChanged(ExcludeNewerValueChange),
+}
+
+impl PackageExcludeNewerChange {
+    pub fn is_relative_timestamp_change(&self) -> bool {
+        match self {
+            Self::Disabled { .. } | Self::Enabled { .. } => false,
+            Self::TimestampChanged(change) => change.is_relative_timestamp_change(),
+        }
+    }
+}
+
+impl std::fmt::Display for PackageExcludeNewerChange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Disabled { was } => {
+                write!(f, "disable exclude newer (was `{was}`)")
+            }
```

Comment:
Should this match the other message, i.e., with "addition of exclude newer exception"?